### ISSUE TITLE
Problem exporting and re-importing prime256v1 EC keys in OpenSSL 3.0

### ISSUE THREAD

**Issue by ddjonin** [openssl/openssl#20444]
I am trying to generate EC EVP_PKEY ```pkey```, export it into the raw uncompressed form with 33 bytes and re-import it to EVP_PKEY ```pkey_out``` on the peer side. While re-importing, it keeps failing with: ```elliptic curve routines:ossl_ec_GFp_simple_oct2point:invalid encoding:../../openssl-3.0.7/crypto/ec/ecp_oct.c:318```

```
    OSSL_PARAM *params;
    OSSL_PARAM_BLD *param_bld = NULL;
    unsigned char* pk = NULL;
    unsigned char buf[34];    
    EVP_PKEY *pkey_out;
    EVP_PKEY_CTX *pkey_ctx;
    static const unsigned char ec_pubkey_prefix[] = { 0x04 }; // Value of POINT_CONVERSION_COMPRESSED from /usr/include/openssl/ec.h
    unsigned char pub[34];
    size_t len;

    /* Export raw public key pub */
    if (!EVP_PKEY_get_octet_string_param(pkey, OSSL_PKEY_PARAM_PUB_KEY,
                                        pub, sizeof(pub), &len)) {
                 /* Error */
    }

    memcpy(buf, ec_pubkey_prefix, 1);
    memcpy(buf + 1, pub, 33);

    /* Build peer EVP_PKEY from 33 byte uncompressed raw public key */
    param_bld = OSSL_PARAM_BLD_new();

    if (param_bld
     && OSSL_PARAM_BLD_push_utf8_string(param_bld, "group", "prime256v1", 0) > 0
     && OSSL_PARAM_BLD_push_octet_string(param_bld, "pub", buf, sizeof(buf)) > 0)
     params = OSSL_PARAM_BLD_to_param(param_bld);

    pkey_ctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL);
    if (pkey_ctx
     && EVP_PKEY_fromdata_init(pkey_ctx) > 0
     && EVP_PKEY_fromdata(pkey_ctx, &pkey_out, EVP_PKEY_PUBLIC_KEY, params) > 0) {
        // pkey initialized successfully
     } else {
        // pkey initialization failed
        openssl_print_error();
     }
```

This keeps failing in ```EVP_PKEY_fromdata```. Is there some obvious problem with the above code that I do not see or it is some underlying OpenSSL3.0 issue?

Unfortunately, I could not find a full example in openssl.org to base this code on.

For reference, the origin ```pkey``` is generatad based on the example in openssl.org wiki ```https://wiki.openssl.org/index.php/Elliptic_Curve_Diffie_Hellman``` with curve ```NID_X9_62_prime256v1```.

**Comment by mattcaswell** [comment]
Refer to the following man page for how `OSSL_PKEY_PARAM_PUB_KEY` is treated with an EC key:

https://www.openssl.org/docs/man3.0/man7/EVP_PKEY-EC.html

In particular:

> The public key value in encoded EC point format conforming to Sec. 2.3.3 and 2.3.4 of the SECG SEC 1 ("Elliptic Curve Cryptography") standard. This parameter is used when importing or exporting the public key value with the EVP_PKEY_fromdata() and EVP_PKEY_todata() functions.
> 
> Note, in particular, that the choice of point compression format used for encoding the exported value via EVP_PKEY_todata() depends on the underlying provider implementation. Before OpenSSL 3.0.8, the implementation of providers included with OpenSSL always opted for an encoding in compressed format, unconditionally. Since OpenSSL 3.0.8, the implementation has been changed to honor the OSSL_PKEY_PARAM_EC_POINT_CONVERSION_FORMAT parameter, if set, or to default to uncompressed format.

This tells you that when getting an EC key via `OSSL_PKEY_PARAM_PUB_KEY` it is already in the SECG SEC 1 format - meaning that it has a leading byte indicating the point compression format. You should not modify it to simply add "0x04" (for uncompressed) on to the front. That is not going to work. It already has the relevant byte.

The second para tells you that in <= 3.0.7 you always got compressed format and you can't change it. This was considered an error and was fixed in 3.0.8. From 3.0.8 onwards it defaults to uncompressed but can be overridden by the setting for `OSSL_PKEY_PARAM_EC_POINT_CONVERSION_FORMAT`.

You don't explicitly say, but the error message suggests that you are using 3.0.7. So, I would recommend that you upgrade to 3.0.8. After upgrading, don't try and modify the returned value. You may want to explicitly set `OSSL_PKEY_PARAM_EC_POINT_CONVERSION_FORMAT` to always force uncompressed format.

> uncompressed form with 33 bytes

Uncompressed prime256v1 keys are not 33 bytes in length

**Comment by ddjonin** [comment]
Thank you. This was helpful.
Regarding ```uncompressed form with 33 bytes``` this was a typo. Should have been ```compressed form with 33 bytes```.
